### PR TITLE
Body Graph UI

### PR DESCRIPTION
**Pull request by dseguin**
#### Summary
None

*Alternative summary: "Data driven body graphs"*

#### Purpose of change
I really wanted to create some infrastructure for creating cool interactive body diagrams in ascii art. This includes a new UI for viewing body part status info.

#### Describe the solution

https://user-images.githubusercontent.com/12537966/165710055-a9f7dc55-b177-4884-bf61-b579433dcee0.mp4

<hr>

##### JSON data

I've added a new JSON data type `"body_graph"`:

```JSON
 {
    "type": "body_graph",
    "id": "torso",
    "parent_bodypart": "torso",
    "fill_sym": "#",
    "fill_color": "white",
    "rows": [
      "                                        ",
      "          22222                22222    ",
      "         2222222                22222   ",
      "    11111122222111111          22###55  ",
      "  ###111111111111111###       44#####55 ",
      " ####111111111111111####     44######55 ",
      "     111111111111111         44######55 ",
      "      1111111111111          44######55 ",
      "      1111111111111           44####55  ",
      "      1111111111111           44####55  ",
      "      1133333333311           44####55  ",
      "      6333333333336           44666655  ",
      "      6333333333336           46666665  ",
      "     663333333333366          66666666  ",
      "     666333333333666          6666666## ",
      "    ######33333######        ########## ",
      "    ########3########        ########## ",
      "    ######## ########        #########  ",
      "    #######   #######         #######   ",
      "                                        "
    ],
    "parts": {
      "1": { "sub_body_parts": [ "torso_upper" ], "select_color": "red" },
      "2": { "sub_body_parts": [ "torso_neck" ], "select_color": "red" },
      "3": { "sub_body_parts": [ "torso_lower" ], "select_color": "red" },
      "4": { "sub_body_parts": [ "torso_hanging_front" ], "select_color": "red" },
      "5": { "sub_body_parts": [ "torso_hanging_back" ], "select_color": "red" },
      "6": { "sub_body_parts": [ "torso_waist" ], "select_color": "red" }
    }
  }
```

Each entry in `"parts"` can also define a `"nested_graph"` field which points to another body_graph, so when the player selects that part they are taken to the nested graph.

See [the updated documentation](https://github.com/CleverRaven/Cataclysm-DDA/blob/23365c6ac7111d79645deeb2a8cb6fbdb478a948/doc/JSON_INFO.md#body-graphs) for more details about these fields.

<hr>

##### Body Graph UI

I created a new UI accessible from the `@` menu (uses the `s` hotkey by default):

hand_graph

This UI shows the player's body in graph form using the body_graph objects. Each main body part has a nested graph that leads to a detailed view of its sub parts.

Currently the UI is used to show equipment stats for that body part in the right pane. (The right info pane was only possible thanks to @bombasticSlacks!)

<hr>

#### Describe alternatives you've considered

Stay in the past, with boring text.

#### Testing

See video and screenshots.

#### Additional context

TODO:
- [x] Add infrastructure for reading body_graph objects
- [x] Create body graphs:
  - [x] Full body
  - [x] Head
  - [x] Torso
  - [x] Arms
  - [x] Hands
  - [x] Legs
  - [x] Feet
- [x] Create basic UI to list body parts and draw the body graphs
- [x] Add infrastructure to simplify displaying equipment stats (See https://github.com/dseguin/Cataclysm-DDA/pull/1, thanks @bombasticSlacks!)
- [x] Add mirroring for opposite graph text (thanks @andrei8l!)
- [x] Documentation

### CODE REVIEW

**Comment by bombasticSlacks** [comment]
For body parts (not sub parts) the protection values are an average. Might be a decent idea to add "average" just for them.

**Comment by andrei8l** [comment]
You can mirror the ASCII art at runtime instead of re-drawing it in JSON
<details>
<summary>quick and dirty deferred POC</summary>

```diff
diff --git a/data/json/bodypart_graphs/arms.json b/data/json/bodypart_graphs/arms.json
index 25fab8d52e..f0067f8415 100644
--- a/data/json/bodypart_graphs/arms.json
+++ b/data/json/bodypart_graphs/arms.json
@@ -5,28 +5,7 @@
     "parent_bodypart": "arm_r",
     "fill_sym": "#",
     "fill_color": "white",
-    "rows": [
-      "                                        ",
-      "                                 11111  ",
-      "                              11111111  ",
-      "                             111111111  ",
-      "                             111111111  ",
-      "                             211111111  ",
-      "                            2222221111  ",
-      "                            2222222211  ",
-      "    444                    2222222222   ",
-      "   444444                 2222222222    ",
-      "    444444444            22222222222    ",
-      "       444444444444     22222222222     ",
-      "         4444444444444422222222222      ",
-      "            444444444442222222222       ",
-      "              444444443333322222        ",
-      "                  444333333322          ",
-      "                     3333333            ",
-      "                        33              ",
-      "                                        ",
-      "                                        "
-    ],
+    "mirror" : "arm_l",
     "parts": {
       "1": { "sub_body_parts": [ "arm_shoulder_r" ], "select_color": "red" },
       "2": { "sub_body_parts": [ "arm_upper_r" ], "select_color": "red" },
diff --git a/src/bodygraph.cpp b/src/bodygraph.cpp
index af9763f90f..462b5bbb11 100644
--- a/src/bodygraph.cpp
+++ b/src/bodygraph.cpp
@@ -80,6 +80,8 @@ void bodygraph::load( const JsonObject &jo, const std::string & )
                 rows.emplace_back( row );
             }
         }
+    } else if( !was_loaded || jo.has_string( "mirror" ) ) {
+        mirror = jo.get_string( "mirror" );
     }
 
     if( !was_loaded || jo.has_object( "parts" ) ) {
@@ -344,6 +346,15 @@ void bodygraph_display::draw_graph()
 {
     werase( w_graph );
     const bodygraph_part *selected_graph = std::get<2>( partlist[sel_part] );
+
+    auto const &all = bodygraph_full_body->get_all();
+    bool const mirror = !id->mirror.empty();
+    auto const it =
+        !mirror ? all.cend()
+                : std::find_if( all.cbegin(), all.cend(), [&]( decltype( *all.cbegin() ) &el ) {
+                      return el.id->id.str() == id->mirror;
+                  } );
+
     std::string selected_sym;
     if( !!selected_graph ) {
         for( const auto &bgp : id->parts ) {
@@ -352,22 +363,27 @@ void bodygraph_display::draw_graph()
             }
         }
     }
-    for( unsigned i = 0; i < id->rows.size() && i < BPGRAPH_MAXROWS; i++ ) {
-        for( unsigned j = 0; j < id->rows[i].size() && j < BPGRAPH_MAXCOLS; j++ ) {
+    decltype( id->rows ) const &rows = mirror ? it->id->rows : id->rows;
+    int i = 0;
+    for( auto const &row : rows ) {
+        int x = mirror ? row.size() - 1 : 0;
+        for( auto const &col : row ) {
             std::string sym = id->fill_sym;
-            nc_color col = id->fill_color;
-            auto iter = id->parts.find( id->rows[i][j] );
+            nc_color color = id->fill_color;
+            auto  iter = id->parts.find( col );
             if( iter != id->parts.end() ) {
                 sym = iter->second.sym;
             }
-            if( id->rows[i][j] == " " ) {
-                col = c_unset;
+            if( col == " " ) {
+                color = c_unset;
                 sym = " ";
-            } else if( id->rows[i][j] == selected_sym ) {
-                col = id->parts.at( selected_sym ).sel_color;
+            } else if( col == selected_sym ) {
+                color = id->parts.at( selected_sym ).sel_color;
             }
-            mvwputch( w_graph, point( j, i ), col, sym );
+            mvwputch( w_graph, point( x, i ), color, sym );
+            x = mirror ? x - 1 : x + 1;
         }
+        i++;
     }
     wnoutrefresh( w_graph );
 }
```
</details>

Many other transformations can be done the same way.

File: src/bodygraph.cpp [hunk 1]
```diff
@@ -0,0 +1,577 @@
...
+            debugmsg( "body_graph \"%s\" defines rows with different widths.", id.c_str() );
+            w_warned = true;
+        } else if( !w_warned && w > BPGRAPH_MAXCOLS ) {
+            debugmsg( "body_graph \"%s\" defines rows with more columns than the maximum (%d).", id.c_str(),
+                      BPGRAPH_MAXCOLS );
+            w_warned = true;
+        }
+        for( int i = ( BPGRAPH_MAXCOLS - w ) / 2; i > 0; i-- ) {
+            r.insert( r.begin(), " " );
+        }
```
Comment (andrei8l) [inline src/bodygraph.cpp:128]:
```suggestion
        r.insert( r.begin(), ( BPGRAPH_MAXCOLS - w ) / 2, " " );
```
same for the other loop-insert blocks

File: src/bodygraph.cpp [hunk 2]
```diff
@@ -0,0 +1,577 @@
...
+
+bodygraph_info::bodygraph_info()
+{
+    std::vector<std::string> worn_names = {};
+    avg_coverage = 0;
+    total_encumbrance = 0;
+    worst_case = resistances();
+    median_case = resistances();
+    best_case = resistances();
+}
```
Comment (andrei8l) [inline src/bodygraph.cpp:195]: `avg_coverage` and `total_encumbrance` can be initialized in the struct declaration and the rest of the block is redundant
```suggestion
```

File: src/bodygraph.cpp [hunk 3]
```diff
@@ -0,0 +1,577 @@
...
+        for( const JsonValue jval : jo.get_array( "rows" ) ) {
+            if( !jval.test_string() ) {
+                jval.throw_error( "\"rows\" array must contain string values." );
+            } else {
+                std::wstring wrow = utf8_to_wstr( jval.get_string() );
+                std::vector<std::string> row;
+                for( const wchar_t wc : wrow ) {
+                    row.emplace_back( wstr_to_utf8( std::wstring( 1, wc ) ) );
+                }
+                rows.emplace_back( row );
```
Comment (BrettDong) [inline src/bodygraph.cpp:80]: If you want to extract individual wide characters from the UTF-8 encoded string, there is a `utf8_display_split` function available.

**Comment by Shodan14** [comment]
Going to be really fun drawing all the mutant parts.

**Comment by Theawesomeboophis** [comment]
This is absolutely incredible, can't wait for this, love the ascii art.

File: src/bodygraph.cpp [hunk 4]
```diff
@@ -0,0 +1,577 @@
...
+            draw_info();
+        } );
+    }
+    return current_ui;
+}
+
+void bodygraph_display::display()
+{
+    shared_ptr_fast<ui_adaptor> current_ui = create_or_get_ui_adaptor();
+    init_ui_windows();
```
[the remark this answers is not in the document]
Comment (dseguin) [reply src/bodygraph.cpp:525]: This one is actually necessary to determine the window width and text padding in the first call to `bodygraph_display::prepare_infotext`

File: src/bodygraph.cpp [hunk 5]
```diff
@@ -0,0 +1,577 @@
...
+
+void bodygraph::load( const JsonObject &jo, const std::string & )
+{
+    optional( jo, was_loaded, "parent_bodypart", parent_bp );
+    optional( jo, was_loaded, "fill_sym", fill_sym );
+    if( jo.has_string( "fill_color" ) ) {
+        fill_color = color_from_string( jo.get_string( "fill_color" ) );
+    }
+
+    if( !was_loaded || jo.has_array( "rows" ) ) {
```
Comment (andrei8l) [inline src/bodygraph.cpp:69]: It doesn't look like you're setting `was_loaded` anywhere and UBSan complains about it too:
```gdb
../cataclysm-dda-git/src/bodygraph.cpp:69:10: runtime error: load of value 255, which is not a valid value for type 'bool'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../cataclysm-dda-git/src/bodygraph.cpp:69:10 in
```

Comment (dseguin) [reply src/bodygraph.cpp:69]: Cheers, I forgot to initialize it in bodygraph.h

File: src/bodygraph.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Qrox) [reply src/bodygraph.cpp:515]: If `prepare_infotext` is dependent on the screen size, it should be called within the resize callback to ensure the data are up to date when the screen is resized while waiting for input.

File: src/bodygraph.cpp [hunk 6]
```diff
@@ -0,0 +1,567 @@
...
+        if( !!std::get<1>( a ) && !!std::get<1>( b ) ) {
+            return std::get<1>( a )->name.translated_lt( std::get<1>( b )->name );
+        }
+        if( std::get<0>( a )->name.translated_ne( std::get<0>( b )->name ) ) {
+            return std::get<0>( a )->name.translated_lt( std::get<0>( b )->name );
+        }
+        if( !std::get<1>( a ) != !std::get<1>( b ) ) {
+            return !std::get<1>( a );
+        }
+        return true;
```
Comment (andrei8l) [inline src/bodygraph.cpp:403]:
```suggestion
        return a != b;
```
to satisfy this requirement
`For all a, comp(a,a)==false`
and fix this asan error:
<details>
<summary>asan.log</summary>

```gdb
==1308027==ERROR: AddressSanitizer: container-overflow on address 0x613000f2e0f8 at pc 0x5582eab4374c bp 0x7ffe8da2c230 sp 0x7ffe8da2c228
READ of size 8 at 0x613000f2e0f8 thread T0
    #0 0x5582eab4374b in bodygraph_display::prepare_partlist()::$_2::operator()(std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*> const&, std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*> const&) const /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/bodygraph.cpp:394:15
    #1 0x5582eab42a10 in void std::__1::__sort<bodygraph_display::prepare_partlist()::$_2&, std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*>*>(std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*>*, std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*>*, bodygraph_display::prepare_partlist()::$_2&) /usr/bin/../include/c++/v1/__algorithm/sort.h:403:24
    #2 0x5582eab3d4ae in void std::__1::sort<std::__1::__wrap_iter<std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*>*>, bodygraph_display::prepare_partlist()::$_2>(std::__1::__wrap_iter<std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*>*>, std::__1::__wrap_iter<std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*>*>, bodygraph_display::prepare_partlist()::$_2) /usr/bin/../include/c++/v1/__algorithm/sort.h:514:9
    #3 0x5582eab3d4ae in bodygraph_display::prepare_partlist() /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/bodygraph.cpp:391:5
    #4 0x5582eab41545 in bodygraph_display::display() /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/bodygraph.cpp:516:5
    #5 0x5582eab42357 in display_bodygraph(string_id<bodygraph> const&) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/bodygraph.cpp:566:9
    #6 0x5582ecf5e38d in handle_player_display_action(Character&, unsigned int&, unsigned int&, (anonymous namespace)::player_display_tab&, input_context&, ui_adaptor const&, ui_adaptor const&, ui_adaptor const&, ui_adaptor const&, ui_adaptor const&, ui_adaptor const&, ui_adaptor const&, ui_adaptor const&, ui_adaptor const&, std::__1::vector<string_id<mutation_branch>, std::__1::allocator<string_id<mutation_branch> > > const&, std::__1::vector<bionic, std::__1::allocator<bionic> > const&, std::__1::vector<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > > const&, std::__1::vector<HeaderSkill, std::__1::allocator<HeaderSkill> > const&, bool) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/player_display.cpp:1131:9
    #7 0x5582ecf59d61 in Character::disp_info(bool) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/player_display.cpp:1598:16
    #8 0x5582ea8bcd24 in game::do_regular_action(action_id&, avatar&, cata::optional<tripoint> const&) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/handle_action.cpp:2472:30
    #9 0x5582ea8eb2b2 in game::handle_action() /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/handle_action.cpp:2893:14
    #10 0x5582e9c908cb in do_turn() /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/do_turn.cpp:702:24
    #11 0x5582e8be1404 in main /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/main.cpp:766:17
    #12 0x7f15a532330f in __libc_start_call_main libc-start.c
    #13 0x7f15a53233c0 in __libc_start_main@GLIBC_2.2.5 (/usr/lib/libc.so.6+0x2d3c0)
    #14 0x5582e8aeb024 in _start (/usr/src/cataclysm-dda-git/src/build_asan/cataclysm-tiles+0x7471024)

0x613000f2e0f8 is located 248 bytes inside of 384-byte region [0x613000f2e000,0x613000f2e180)
allocated by thread T0 here:
    #0 0x5582e8bdd4f1 in operator new(unsigned long) (/usr/src/cataclysm-dda-git/src/build_asan/cataclysm-tiles+0x75634f1)
    #1 0x5582eab6a3a6 in void* std::__1::__libcpp_operator_new<unsigned long>(unsigned long) /usr/bin/../include/c++/v1/new:235:10
    #2 0x5582eab6a3a6 in std::__1::__libcpp_allocate(unsigned long, unsigned long) /usr/bin/../include/c++/v1/new:261:10
    #3 0x5582eab6a3a6 in std::__1::allocator<std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*> >::allocate(unsigned long) /usr/bin/../include/c++/v1/__memory/allocator.h:108:38
    #4 0x5582eab6a3a6 in std::__1::allocator_traits<std::__1::allocator<std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*> > >::allocate(std::__1::allocator<std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*> >&, unsigned long) /usr/bin/../include/c++/v1/__memory/allocator_traits.h:262:20
    #5 0x5582eab6a3a6 in std::__1::__split_buffer<std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*>, std::__1::allocator<std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*> >&>::__split_buffer(unsigned long, unsigned long, std::__1::allocator<std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*> >&) /usr/bin/../include/c++/v1/__split_buffer:315:29
    #6 0x5582eab69a7f in void std::__1::vector<std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*>, std::__1::allocator<std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*> > >::__emplace_back_slow_path<std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*> >(std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*>&&) /usr/bin/../include/c++/v1/vector:1687:49
    #7 0x5582eab3d393 in bodygraph_display::prepare_partlist() /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/bodygraph.cpp:384:22
    #8 0x5582eab41545 in bodygraph_display::display() /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/bodygraph.cpp:516:5
    #9 0x5582eab42357 in display_bodygraph(string_id<bodygraph> const&) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/bodygraph.cpp:566:9
    #10 0x5582ecf5e38d in handle_player_display_action(Character&, unsigned int&, unsigned int&, (anonymous namespace)::player_display_tab&, input_context&, ui_adaptor const&, ui_adaptor const&, ui_adaptor const&, ui_adaptor const&, ui_adaptor const&, ui_adaptor const&, ui_adaptor const&, ui_adaptor const&, ui_adaptor const&, std::__1::vector<string_id<mutation_branch>, std::__1::allocator<string_id<mutation_branch> > > const&, std::__1::vector<bionic, std::__1::allocator<bionic> > const&, std::__1::vector<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > > const&, std::__1::vector<HeaderSkill, std::__1::allocator<HeaderSkill> > const&, bool) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/player_display.cpp:1131:9
    #11 0x5582ecf59d61 in Character::disp_info(bool) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/player_display.cpp:1598:16
    #12 0x5582ea8bcd24 in game::do_regular_action(action_id&, avatar&, cata::optional<tripoint> const&) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/handle_action.cpp:2472:30
    #13 0x5582ea8eb2b2 in game::handle_action() /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/handle_action.cpp:2893:14
    #14 0x5582e9c908cb in do_turn() /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/do_turn.cpp:702:24
    #15 0x5582e8be1404 in main /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/main.cpp:766:17
    #16 0x7f15a532330f in __libc_start_call_main libc-start.c

HINT: if you don't care about these errors you may set ASAN_OPTIONS=detect_container_overflow=0.
If you suspect a false positive see also: https://github.com/google/sanitizers/wiki/AddressSanitizerContainerOverflow.
SUMMARY: AddressSanitizer: container-overflow /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/bodygraph.cpp:394:15 in bodygraph_display::prepare_partlist()::$_2::operator()(std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*> const&, std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*> const&) const
Shadow bytes around the buggy address:
  0x0c26801ddbc0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c26801ddbd0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c26801ddbe0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c26801ddbf0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c26801ddc00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x0c26801ddc10: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 fc[fc]
  0x0c26801ddc20: fc fc fc fc fc fc fc fc fc fc fc fc fc fc fc fc
  0x0c26801ddc30: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c26801ddc40: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c26801ddc50: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c26801ddc60: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==1308027==ABORTING
```
</details>

Comment (dseguin) [reply src/bodygraph.cpp:403]: Good find, appreciated!

File: src/bodygraph.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (dseguin) [reply src/bodygraph.cpp]: That did the trick, cheers!

**Comment by perryprog** [comment]
I'm wondering if there should be a merger (or at least unification) between this and the medical menu introduced in #54763 due to a fair bit of overlap between the two; in my opinion, it doesn't really make sense to keep this as two separate displays. (Reasons against merging them could be just down to personally preference on wanting / not wanting a large ascii art diagram for whatever a player is currently doing.)

I also wonder if the ascii art should be hidden or represented in text based on if `ENABLE_ASCII_ART` is set.

**Comment by I-am-Erk** [comment]
Almost certainly the two will be integrated, but it might wait until after the wounds overhaul in 0.H-experimental.

**Pull request merged**